src/core/iomgr/reachability.h

```diff
@@ -0,0 +1,93 @@
+/*
+ *
+ * Copyright 2015, Google Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#ifndef GRPC_INTERNAL_CORE_IOMGR_REACHABILITY_H
+#define GRPC_INTERNAL_CORE_IOMGR_REACHABILITY_H
+
+/// Network type: a broad classification of different kinds of network
+/// Useful for systems that want to send traffic only on a restricted class
+typedef enum grpc_network_type {
+  /// Network type is unknown
+  GRPC_NETWORK_UNKNOWN_TYPE,
+  /// Network is a wired connection
+  GRPC_NETWORK_WIRED,
+  /// Network is a wifi connection
+  GRPC_NETWORK_WIFI,
+  /// Network is a cellular connection
+  GRPC_NETWORK_CELLULAR
+} grpc_network_type;
+
+/// A description of a nic
+typedef struct grpc_network_interface {
+  /// Type of the network this nic refers to
+  grpc_network_type network_type;
+  /// Address to pass to bind() to select this nic
+  const sockaddr_t *bind_addr;
+  size_t bind_addr_len;
+  /// A name a sysadmin could use to identify this nic
+  const char *name;
+} grpc_network_interface;
+
+/// Copy a network interface description
+grpc_network_interface *grpc_network_interface_copy(
+    grpc_network_interface *src);
+/// Destroy a network interface description
+void grpc_network_interface_destroy(grpc_network_interface *nic);
```

Comment:
We'll either need to be able to pass around (and copy) these descriptors, or reference count them. (imagine running ifup/ifdown equivalents on a server: the set of nics can change dynamically, so pre-allocating won't be enough).

Copying seems easier to localize bugs with, and this isn't critical path, so I'm preferring copying.